Hipótesis principal: Los comentarios de los clientes contienen suficiente información textual que puede ser utilizada para predecir de manera precisa las categorías de satisfacción o problemas específicos, como la calidad de las instalaciones, el servicio al cliente, etc.

In [105]:
#Etiquetar los comentarios

import pandas as pd

df_sentimientos_hotel=pd.read_csv('.\CHATBOT\hotel_sentimientos_general.csv')
df_sentimientos_hotel

Unnamed: 0.1  Unnamed: 0   
0                 0           0  \
1                 1           1   
2                 2           2   
3                 3           3   
4                 4           4   
...             ...         ...   
13127         13127       13127   
13128         13128       13128   
13129         13129       13129   
13130         13130       13130   
13131         13131       13131   

                                       Translated_Review  Rating Sentimiento   
0      bonito hotel, aparcamiento caro, consiguió una...       4    POSITIVO  \
1      ok, nada de cargo especial, miembro diamante h...       2    POSITIVO   
2      Bonitas habitaciones, no un hotel de 4* con ex...       3    NEGATIVO   
3      Único, estancia estupenda, tiempo maravilloso ...       5     NEUTRAL   
4      estancia estupenda estancia estupenda, fue inc...       5     NEUTRAL   
...                                                  ...     ...         ...   
13127  Los viajeros caribeños más agradables y experi...       5    POSITIVO   
13128  maravilloso club de vacaciones diamante bahia ...       5    POSITIVO   
13129  No me molesté en quedarme en la puerta, me que...       1    NEGATIVO   
13130  escribió el encabezado, breve. comida repugnan...       1     NEUTRAL   
13131  excelente piscina, clima increíble, estadía fa...       3    POSITIVO   

      Sentimiento_VADER SENTIMIENTO_CALCULADO  
0              NEGATIVO              POSITIVO  
1              NEGATIVO               NEUTRAL  
2              NEGATIVO              NEGATIVO  
3              NEGATIVO              POSITIVO  
4              NEGATIVO              POSITIVO  
...                 ...                   ...  
13127          NEGATIVO              POSITIVO  
13128          NEGATIVO              POSITIVO  
13129          NEGATIVO              NEGATIVO  
13130           NEUTRAL              NEGATIVO  
13131          POSITIVO              POSITIVO  

[13132 rows x 7 columns]

In [7]:
#filtar los comentarios negativos para identificar el problema

df_comentarios_negativos=df_sentimientos_hotel[df_sentimientos_hotel['SENTIMIENTO_CALCULADO']=='NEGATIVO']
df_comentarios_negativos

Unnamed: 0.1  Unnamed: 0   
2                 2           2  \
15               15          15   
19               19          19   
32               32          32   
40               40          40   
...             ...         ...   
13070         13070       13070   
13101         13101       13101   
13116         13116       13116   
13129         13129       13129   
13130         13130       13130   

                                       Translated_Review  Rating Sentimiento   
2      Bonitas habitaciones, no un hotel de 4* con ex...       3    NEGATIVO  \
15     horrible servicio al cliente estancia en el ho...       1    NEGATIVO   
19     hmmmmm, dice que hay muchas esperanzas en el h...       3    NEGATIVO   
32     Ruido, aire acondicionado: estándar, estadía o...       1     NEUTRAL   
40     Mala elección, cable caliente del hotel reserv...       1     NEUTRAL   
...                                                  ...     ...         ...   
13070  Malo, poco acogedor, mal servicio, habitación ...       1     NEUTRAL   
13101  buen hotel con un presupuesto pequeño nos hosp...       4    NEGATIVO   
13116  no es el secreto mejor guardado, promedio mejo...       2    NEGATIVO   
13129  No me molesté en quedarme en la puerta, me que...       1    NEGATIVO   
13130  escribió el encabezado, breve. comida repugnan...       1     NEUTRAL   

      Sentimiento_VADER SENTIMIENTO_CALCULADO  
2              NEGATIVO              NEGATIVO  
15             NEGATIVO              NEGATIVO  
19             POSITIVO              NEGATIVO  
32             NEGATIVO              NEGATIVO  
40             NEGATIVO              NEGATIVO  
...                 ...                   ...  
13070          NEGATIVO              NEGATIVO  
13101          NEGATIVO              NEGATIVO  
13116          NEGATIVO              NEGATIVO  
13129          NEGATIVO              NEGATIVO  
13130           NEUTRAL              NEGATIVO  

[1534 rows x 7 columns]

In [109]:

#Carga de los csv del top 3 de hoteles con sentimientos negativos:     Fairmont,    Hyatt,    Homewood

df_sentimientos_competencia=pd.read_csv('.\df_hoteles_Competencia.csv')
df_sentimientos_competencia

Unnamed: 0 Nombre Hotel   
0             0      affinia  \
1             1      affinia   
2             2      affinia   
3             3      affinia   
4             4      affinia   
..          ...          ...   
795         795      talbott   
796         796      talbott   
797         797      talbott   
798         798      talbott   
799         799      talbott   

                                            Comentario Sentimiento   
0    Después de una estancia de una semana reciente...     NEUTRAL  \
1    Aunque demasiado caro en mi opinión, el hotel ...     NEUTRAL   
2    Me alojé en el Affina Chicago para celebrar mi...     NEUTRAL   
3    ESTE HOTEL ES FANTÁSTICO. Me alojé allí en mi ...     NEUTRAL   
4    El Affinia Chicago es un lugar maravilloso don...     NEUTRAL   
..                                                 ...         ...   
795  Me alojé en The Talbott por 3 noches por negoc...     NEUTRAL   
796  Nos encantó el Talbott. La ubicación era fabul...     NEUTRAL   
797  Puedo decir fácilmente que este es uno de los ...    NEGATIVO   
798  Este es un gran hallazgo para el centro de Chi...    POSITIVO   
799  Chicago es una de nuestras ciudades favoritas ...     NEUTRAL   

    Sentimiento_VADER  
0            POSITIVO  
1            POSITIVO  
2            POSITIVO  
3            POSITIVO  
4            POSITIVO  
..                ...  
795          POSITIVO  
796          POSITIVO  
797          POSITIVO  
798          POSITIVO  
799           NEUTRAL  

[800 rows x 5 columns]

In [78]:
df_sentimientos_competencia_positivos=df_sentimientos_competencia[df_sentimientos_competencia['Sentimiento_VADER']=='POSITIVO']
df_sentimientos_competencia_positivos

Unnamed: 0 Nombre Hotel   
0             0      affinia  \
1             1      affinia   
2             2      affinia   
3             3      affinia   
4             4      affinia   
..          ...          ...   
792         792      talbott   
795         795      talbott   
796         796      talbott   
797         797      talbott   
798         798      talbott   

                                            Comentario Sentimiento   
0    Después de una estancia de una semana reciente...     NEUTRAL  \
1    Aunque demasiado caro en mi opinión, el hotel ...     NEUTRAL   
2    Me alojé en el Affina Chicago para celebrar mi...     NEUTRAL   
3    ESTE HOTEL ES FANTÁSTICO. Me alojé allí en mi ...     NEUTRAL   
4    El Affinia Chicago es un lugar maravilloso don...     NEUTRAL   
..                                                 ...         ...   
792  Cualquier viajero fiel a su hotel favorito ha ...     NEUTRAL   
795  Me alojé en The Talbott por 3 noches por negoc...     NEUTRAL   
796  Nos encantó el Talbott. La ubicación era fabul...     NEUTRAL   
797  Puedo decir fácilmente que este es uno de los ...    NEGATIVO   
798  Este es un gran hallazgo para el centro de Chi...    POSITIVO   

    Sentimiento_VADER  
0            POSITIVO  
1            POSITIVO  
2            POSITIVO  
3            POSITIVO  
4            POSITIVO  
..                ...  
792          POSITIVO  
795          POSITIVO  
796          POSITIVO  
797          POSITIVO  
798          POSITIVO  

[479 rows x 5 columns]

In [93]:
df_sentimientos_competencia_positivostop=df_sentimientos_competencia_positivos[(df_sentimientos_competencia_positivos['Nombre Hotel']=='knickerboxer')|(df_sentimientos_competencia_positivos['Nombre Hotel']=='talbott')|(df_sentimientos_competencia_positivos['Nombre Hotel']=='sofitel')]
df_sentimientos_competencia_positivostop

Unnamed: 0 Nombre Hotel   
340         340      sofitel  \
342         342      sofitel   
343         343      sofitel   
344         344      sofitel   
345         345      sofitel   
347         347      sofitel   
350         350      sofitel   
355         355      sofitel   
359         359      sofitel   
380         380      talbott   
381         381      talbott   
382         382      talbott   
383         383      talbott   
384         384      talbott   
385         385      talbott   
386         386      talbott   
387         387      talbott   
388         388      talbott   
389         389      talbott   
390         390      talbott   
391         391      talbott   
392         392      talbott   
393         393      talbott   
394         394      talbott   
395         395      talbott   
397         397      talbott   
399         399      talbott   
741         741      sofitel   
742         742      sofitel   
743         743      sofitel   
745         745      sofitel   
747         747      sofitel   
748         748      sofitel   
749         749      sofitel   
750         750      sofitel   
751         751      sofitel   
753         753      sofitel   
754         754      sofitel   
755         755      sofitel   
756         756      sofitel   
757         757      sofitel   
758         758      sofitel   
759         759      sofitel   
780         780      talbott   
783         783      talbott   
785         785      talbott   
786         786      talbott   
788         788      talbott   
790         790      talbott   
791         791      talbott   
792         792      talbott   
795         795      talbott   
796         796      talbott   
797         797      talbott   
798         798      talbott   

                                            Comentario Sentimiento   
340  Necesitaba un lugar donde quedarme para una co...    NEGATIVO  \
342  Me alojé en Sofitel con mi marido durante un f...     NEUTRAL   
343  Mi marido y yo nos alojamos en el Sofitel Chic...     NEUTRAL   
344  Alojarme en el Sofitel fue una de las experien...     NEUTRAL   
345  Después de llegar al hotel Sofitel Chicago Wat...     NEUTRAL   
347  El Sofitel Chicago Water Tower, aunque en teor...    POSITIVO   
350  Me alojé en este hotel por 2 noches. Tenía gra...     NEUTRAL   
355  Hace poco me alojé en el hotel Sofitel (Chicag...    POSITIVO   
359  Estaba aquí por negocios, así que necesitaba d...     NEUTRAL   
380  Recientemente me alojé en el Hotel Talbott en ...     NEUTRAL   
381  "Con clase", "Decente" y "Serena". Estas palab...     NEUTRAL   
382  ¡El Hotel Talbott es fantástico! Me quedé allí...     NEUTRAL   
383  Este hotel era todo lo que estábamos buscando....     NEUTRAL   
384  Mi marido y yo decidimos planear una escapada ...     NEUTRAL   
385  Este es el mejor hotel para alojarse cuando vi...     NEUTRAL   
386  Mi estadía en el Talbott fue una experiencia m...     NEUTRAL   
387  ¡Gran hotel! El personal es amable, cortés y p...     NEUTRAL   
388  Realmente me encantó mi estancia en el Talbott...     NEUTRAL   
389  Este hotel fue un auténtico sueño en el que al...     NEUTRAL   
390  Muy lujoso y vale la pena el dinero. Si busca ...     NEUTRAL   
391  Estaba visitando a unos familiares en Chicago ...    POSITIVO   
392  El hotel Talbot es un lugar elegante para llev...     NEUTRAL   
393  El ambiente en el Talbott Hotel es muy acogedo...    POSITIVO   
394  Al principio, mi esposo y yo estábamos muy emo...    POSITIVO   
395  Las tarifas en The Talbott Hotel eran más bara...    POSITIVO   
397  Personal agradable y servicio de limpieza. Des...     NEUTRAL   
399  Excelente personal y servicio al cliente, muy ...     NEUTRAL   
741  Retiré mi Lexus y me lo devolvieron con el esp...     NEUTRAL   
742  Tuve la ceremonia de mi boda y la recepción en...    POSITIVO   
743  Leí las críticas en Trip Advisor sobre esta pr...     NEUTRAL   
745  Acabo de regresar de una visita de 10

In [79]:
df_sentimientos_competencia_negativos=df_sentimientos_competencia[df_sentimientos_competencia['Sentimiento_VADER']=='NEGATIVO']
df_sentimientos_competencia_negativos

Unnamed: 0 Nombre Hotel   
12           12      affinia  \
22           22      allegro   
28           28      allegro   
30           30      allegro   
33           33      allegro   
..          ...          ...   
782         782      talbott   
784         784      talbott   
787         787      talbott   
789         789      talbott   
794         794      talbott   

                                            Comentario Sentimiento   
12   Recientemente me alojé en el hotel Affinia en ...    POSITIVO  \
22   Recientemente visité Chicago. Me alojé en el H...    POSITIVO   
28   Me he alojado en el Allegro durante dos días. ...     NEUTRAL   
30   Su sitio web me cautivó lo suficiente como par...     NEUTRAL   
33   Quedé muy insatisfecho con mi estancia en el A...     NEUTRAL   
..                                                 ...         ...   
782  No soy muy quejoso, como puedes ver al revisar...     NEUTRAL   
784  El Talbott Hotel en sí está un poco desgastado...    NEGATIVO   
787  Ubicación, ubicación, ubicación. ¡Qué gran hal...     NEUTRAL   
789  Había solicitado una habitación tranquila con ...    POSITIVO   
794  Encontré este hotel en el folleto "Pequeños ho...     NEUTRAL   

    Sentimiento_VADER  
12           NEGATIVO  
22           NEGATIVO  
28           NEGATIVO  
30           NEGATIVO  
33           NEGATIVO  
..                ...  
782          NEGATIVO  
784          NEGATIVO  
787          NEGATIVO  
789          NEGATIVO  
794          NEGATIVO  

[229 rows x 5 columns]

In [94]:
df_sentimientos_competencia_negativostop=df_sentimientos_competencia_positivos[(df_sentimientos_competencia_positivos['Nombre Hotel']=='homewood')|(df_sentimientos_competencia_positivos['Nombre Hotel']=='fairmont')|(df_sentimientos_competencia_positivos['Nombre Hotel']=='hyatt')]
df_sentimientos_competencia_negativostop

Unnamed: 0 Nombre Hotel   
100         100     fairmont  \
101         101     fairmont   
103         103     fairmont   
106         106     fairmont   
112         112     fairmont   
114         114     fairmont   
115         115     fairmont   
117         117     fairmont   
163         163     homewood   
166         166     homewood   
168         168     homewood   
169         169     homewood   
171         171     homewood   
172         172     homewood   
174         174     homewood   
176         176     homewood   
178         178     homewood   
180         180        hyatt   
181         181        hyatt   
185         185        hyatt   
187         187        hyatt   
190         190        hyatt   
191         191        hyatt   
192         192        hyatt   
194         194        hyatt   
196         196        hyatt   
198         198        hyatt   
501         501     fairmont   
504         504     fairmont   
505         505     fairmont   
507         507     fairmont   
508         508     fairmont   
509         509     fairmont   
512         512     fairmont   
514         514     fairmont   
516         516     fairmont   
519         519     fairmont   
561         561     homewood   
562         562     homewood   
563         563     homewood   
565         565     homewood   
567         567     homewood   
573         573     homewood   
574         574     homewood   
575         575     homewood   
577         577     homewood   
578         578     homewood   
580         580        hyatt   
583         583        hyatt   
584         584        hyatt   
585         585        hyatt   
590         590        hyatt   
591         591        hyatt   
595         595        hyatt   
597         597        hyatt   
599         599        hyatt   

                                            Comentario Sentimiento   
100  Reservamos un paquete American Girl Place en e...     NEUTRAL  \
101  Me alojé recientemente en el Fairmont Chicago ...     NEUTRAL   
103  El Fairmont Chicago Millennium Park tiene que ...    POSITIVO   
106  Mi estancia en el Fairmont Chicago Millennium ...     NEUTRAL   
112  ¡Mi experiencia con el hotel Fairmont Chicago ...    NEGATIVO   
114  Me hospedé en el Fairmont Chicago Millennium P...     NEUTRAL   
115  Siempre que decidí quedarme en el Fairmont Chi...    POSITIVO   
117  Recientemente me alojé en el hotel Fairmont Ch...    POSITIVO   
163  Me alojé en el Homewood Suites en el centro de...     NEUTRAL   
166  Fui al Homewood Suites en Chicago, que forma p...    POSITIVO   
168  Mi estancia en el Hilton fue como vivir en el ...     NEUTRAL   
169  Al entrar en Homewood Suites, fui recibido por...     NEUTRAL   
171  El Homewood Suites by Hilton, en el centro de ...     NEUTRAL   
172  Recientemente me alojé en el Homewood Suites b...    POSITIVO   
174  Dios mío, este tiene que ser uno de los peores...    POSITIVO   
176  ¡El Homewood Suites es uno de mis hoteles favo...     NEUTRAL   
178  Después de considerar varios hoteles en el áre...    POSITIVO   
180  El hotel Hyatt Regency de Chicago fue el peor ...     NEUTRAL   
181  Durante mi reciente estadía en el Hyatt Regenc...    POSITIVO   
185  Cuando revisé por primera vez el sitio web y l...     NEUTRAL   
187  El Hyatt Regency Chicago parecía un buen lugar...     NEUTRAL   
190  Recientemente me alojé en el hotel Hyatt Regen...     NEUTRAL   
191  Mi esposa y yo fuimos a Chicago para disfrutar...     NEUTRAL   
192  Hotel Hyatt Regency: buen centro de Chicago. L...     NEUTRAL   
194  Mi esposa y yo nos hospedamos en el Hyatt Rege...    NEGATIVO   
196  Decidimos gastar un poco más para alojarnos en...    POSITIVO   
198  Hace poco me alojé en el Hyatt Regency en Chic...     NEUTRAL   
501  Los hoteles Fairmont son generalmente mi opció...    POSITIVO   
504  Tenía muchas ganas de quedarme aquí para mi bo...     NEUTRAL   
505  Hermoso hotel, pero no tiene mucho que ofrecer...     NEUTRAL   
507  El ho

In [61]:
import nltk
from nltk.corpus import stopwords

# Descargar los recursos necesarios de NLTK (esto solo se necesita una vez)
nltk.download('stopwords')

# Definir las categorías de problemas y palabras clave
categorias_problemas = {
    'Precio y costos adicionales': [
        'caro', 'tarifa', 'costo', 'elevado', 'precio', 'presupuesto', 'económico', 'costoso', 'gasto', 
        'inversión', 'incremento', 'subida', 'alza', 'aumento', 'desembolso', 'desembolso', 'cantidad', 
        'nivel', 'monto', 'característica', 'propiedad', 'importe', 'grado', 'escalón', 'tasa', 'pago', 
        'especie', 'coste', 'concepto', 'plus', 'cupo', 'clase', 'derecho', 'cuantía', 'cuota', 'chance', 
        'impuesto', 'precio', 'tiquete', 'grava', 'desglose', 'valla', 'impresión', 'adición', 'deuda', 
        'coste', 'oferta', 'estimación', 'costo', 'valor', 'coste', 'impuesto', 'tasa', 'inflación', 'caída', 
        'preocupación', 'billete', 'cambio', 'interés', 'duda', 'descuento', 'gobierno', 'moneda', 'resto', 
        'distribución', 'compra', 'futuro', 'asociación', 'parte', 'regulación', 'decreto', 'financiación', 
        'subvención', 'planificación', 'periodo', 'economía', 'sistema', 'actuación', 'costo', 'barra', 'tienda', 
        'inmueble', 'financiamiento', 'inversión', 'dinero', 'valoración', 'tarifa', 'distribución', 'prueba', 
        'aplicación', 'importación', 'precio', 'retiro', 'curso', 'alma', 'grano', 'cultura', 'razón', 'carro', 
        'metodología', 'precio','ticket','elevado','carísimo'
    ],
    'Calidad de las instalaciones': [
        'habitación', 'instalación', 'limpieza', 'mantenimiento', 'condiciones', 'estado', 'aseo', 'aseado', 
        'higiene', 'confort', 'infraestructura', 'mejora', 'servicios', 'servicios', 'ambientación', 'instalación', 
        'edificio', 'hospitalidad', 'mobiliario', 'hospedaje', 'comodidad', 'apariencia', 'materiales', 'estructura', 
        'hoteles', 'vivienda', 'diseño', 'construcción', 'decoración', 'comodidades', 'hogar', 'residencia', 'acabados', 
        'lugar', 'inmueble', 'casa', 'estudio', 'vivienda', 'cómoda', 'cuidadoso', 'domicilio', 'detalle', 'operativo', 
        'respeto', 'especialización', 'bien', 'calidad', 'planta', 'tamaño', 'familia', 'altura', 'trabajo', 'recinto', 
        'problema', 'alivio', 'clima', 'proceso', 'oficina', 'sistema', 'juego', 'cuerpo', 'tela', 'edad', 'grupo', 'abuso', 
        'usuario', 'estación', 'ataque', 'perro', 'concepto', 'título', 'fecha', 'rojo', 'zona', 'reserva', 'enfermedad', 
        'núcleo', 'realidad', 'deber', 'compra', 'relación', 'minuto', 'problema', 'boca', 'comunidad', 'norma', 'fuerza', 
        'riesgo', 'cultura', 'tendencia', 'patrón', 'tradición', 'filosofía', 'opción', 'naturaleza', 'representación', 
        'fondo', 'fondo','aire','acondicionado','ruido'
    ],
    'Servicio al cliente': [
        'servicio', 'personal', 'atención', 'amabilidad', 'respuesta', 'trato', 'recepción', 'ayuda', 'cliente', 
        'información', 'soporte', 'disposición', 'satisfacción', 'cortesía', 'consultoría', 'asistencia', 'calidad', 
        'recomendación', 'cordialidad', 'profesionalismo', 'hospitalidad', 'amistad', 'tacto', 'asesoría', 'utilización', 
        'necesidad', 'bienvenida', 'preparación', 'respuesta', 'servidor', 'atmósfera', 'oportunidad', 'uso', 'asistente', 
        'integridad', 'precaución', 'información', 'paciencia', 'vocación', 'voto', 'cuestión', 'dedicación', 'actitud', 
        'problema', 'presencia', 'asistente', 'retiro', 'entrega', 'grosero', 'grosera', 'rudo', 'groseros', 'bruscamente', 
        'altanero', 'servicio', 'cliente', 'plazo', 'disposición', 'práctica', 'servicio', 'prestigio', 'prueba', 'entrenamiento', 
        'especialidad', 'garantía', 'personalidad', 'país', 'riesgo', 'búsqueda', 'pasado', 'posibilidad', 'capítulo', 'profesional', 
        'tratamiento', 'lugar', 'solución', 'actitud', 'familia', 'especie', 'compañía', 'modelo', 'firma', 'materia', 'desarrollo', 
        'centro', 'vida', 'crecimiento', 'casa', 'empresa', 'visión', 'economía', 'movimiento', 'esquema', 'comunicación', 'parte', 
        'comité', 'comunidad', 'dirección', 'esfuerzo', 'momento', 'gobierno', 'término', 'relación', 'ciudad', 'educación', 'empresa'
    ],
   'Experiencia general': [
        'experiencia', 'estancia', 'comida', 'visita', 'viaje', 'impresiones', 'opinión', 'satisfacción', 'comentario', 
        'momento', 'estudio', 'reacción', 'recuerdo', 'instante', 'contemplación', 'comunicación', 'tarea', 'hábito', 
        'desarrollo', 'condición', 'impacto', 'presente', 'concepto', 'institución', 'ver', 'formación', 'compromiso', 
        'pasión', 'carácter', 'especie','general']
}


# Obtener palabras vacías (stop words) para español
stop_words = set(stopwords.words('spanish'))

import nltk
from collections import defaultdict

def etiquetar_problema_nltk(comentario, categorias_problemas, stop_words):
    tokens = nltk.word_tokenize(comentario.lower())
    categoria_count = defaultdict(int)

    for categoria, palabras_clave in categorias_problemas.items():
        for palabra_clave in palabras_clave:
            if palabra_clave in tokens and palabra_clave not in stop_words:
                categoria_count[categoria] += 1

    # Encontrar la categoría con el recuento máximo
    if categoria_count:
        categoria_max = max(categoria_count, key=categoria_count.get)
        return categoria_max
    else:
        return 'Experiencia general'




[nltk_data] Downloading package stopwords to
[nltk_data]     e:\Users\1077730\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [62]:
import nltk
nltk.download('punkt')


[nltk_data] Downloading package punkt to
[nltk_data]     e:\Users\1077730\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [117]:
import nltk
from collections import defaultdict

def etiquetar_problema_nltk(comentario, categorias_problemas, stop_words):
    tokens = nltk.word_tokenize(comentario.lower())
    categoria_count = defaultdict(int)

    for categoria, palabras_clave in categorias_problemas.items():
        for palabra_clave in palabras_clave:
            if palabra_clave in tokens and palabra_clave not in stop_words:
                categoria_count[categoria] += 1

    if categoria_count:
        categoria_max = max(categoria_count, key=categoria_count.get)
        return categoria_max
    else:
        return 'Experiencia general'

# Aplicar la función etiquetar_problema_nltk al DataFrame usando lambda para pasar argumentos adicionales
df_sentimientos_competencia['Categoria_Problema'] = df_sentimientos_competencia['Comentario'].apply(lambda x: etiquetar_problema_nltk(x, categorias_problemas, stop_words))



In [65]:
df_comentarios_negativos

Unnamed: 0.1  Unnamed: 0   
2                 2           2  \
15               15          15   
19               19          19   
32               32          32   
40               40          40   
...             ...         ...   
13070         13070       13070   
13101         13101       13101   
13116         13116       13116   
13129         13129       13129   
13130         13130       13130   

                                       Translated_Review  Rating Sentimiento   
2      Bonitas habitaciones, no un hotel de 4* con ex...       3    NEGATIVO  \
15     horrible servicio al cliente estancia en el ho...       1    NEGATIVO   
19     hmmmmm, dice que hay muchas esperanzas en el h...       3    NEGATIVO   
32     Ruido, aire acondicionado: estándar, estadía o...       1     NEUTRAL   
40     Mala elección, cable caliente del hotel reserv...       1     NEUTRAL   
...                                                  ...     ...         ...   
13070  Malo, poco acogedor, mal servicio, habitación ...       1     NEUTRAL   
13101  buen hotel con un presupuesto pequeño nos hosp...       4    NEGATIVO   
13116  no es el secreto mejor guardado, promedio mejo...       2    NEGATIVO   
13129  No me molesté en quedarme en la puerta, me que...       1    NEGATIVO   
13130  escribió el encabezado, breve. comida repugnan...       1     NEUTRAL   

      Sentimiento_VADER SENTIMIENTO_CALCULADO            Categoria_Problema  
2              NEGATIVO              NEGATIVO   Precio y costos adicionales  
15             NEGATIVO              NEGATIVO           Servicio al cliente  
19             POSITIVO              NEGATIVO           Servicio al cliente  
32             NEGATIVO              NEGATIVO  Calidad de las instalaciones  
40             NEGATIVO              NEGATIVO  Calidad de las instalaciones  
...                 ...                   ...                           ...  
13070          NEGATIVO              NEGATIVO   Precio y costos adicionales  
13101          NEGATIVO              NEGATIVO   Precio y costos adicionales  
13116          NEGATIVO              NEGATIVO   Precio y costos adicionales  
13129          NEGATIVO              NEGATIVO           Servicio al cliente  
13130           NEUTRAL              NEGATIVO  Calidad de las instalaciones  

[1534 rows x 8 columns]

In [96]:
df_sentimientos_competencia_negativostop['Categoria_Problema'] = df_sentimientos_competencia_negativostop['Comentario'].apply(lambda x: etiquetar_problema_nltk(x, categorias_problemas, stop_words))
df_sentimientos_competencia_negativostop

e:\Users\1077730\AppData\Local\Temp\ipykernel_24372\1296745121.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Unnamed: 0 Nombre Hotel   
100         100     fairmont  \
101         101     fairmont   
103         103     fairmont   
106         106     fairmont   
112         112     fairmont   
114         114     fairmont   
115         115     fairmont   
117         117     fairmont   
163         163     homewood   
166         166     homewood   
168         168     homewood   
169         169     homewood   
171         171     homewood   
172         172     homewood   
174         174     homewood   
176         176     homewood   
178         178     homewood   
180         180        hyatt   
181         181        hyatt   
185         185        hyatt   
187         187        hyatt   
190         190        hyatt   
191         191        hyatt   
192         192        hyatt   
194         194        hyatt   
196         196        hyatt   
198         198        hyatt   
501         501     fairmont   
504         504     fairmont   
505         505     fairmont   
507         507     fairmont   
508         508     fairmont   
509         509     fairmont   
512         512     fairmont   
514         514     fairmont   
516         516     fairmont   
519         519     fairmont   
561         561     homewood   
562         562     homewood   
563         563     homewood   
565         565     homewood   
567         567     homewood   
573         573     homewood   
574         574     homewood   
575         575     homewood   
577         577     homewood   
578         578     homewood   
580         580        hyatt   
583         583        hyatt   
584         584        hyatt   
585         585        hyatt   
590         590        hyatt   
591         591        hyatt   
595         595        hyatt   
597         597        hyatt   
599         599        hyatt   

                                            Comentario Sentimiento   
100  Reservamos un paquete American Girl Place en e...     NEUTRAL  \
101  Me alojé recientemente en el Fairmont Chicago ...     NEUTRAL   
103  El Fairmont Chicago Millennium Park tiene que ...    POSITIVO   
106  Mi estancia en el Fairmont Chicago Millennium ...     NEUTRAL   
112  ¡Mi experiencia con el hotel Fairmont Chicago ...    NEGATIVO   
114  Me hospedé en el Fairmont Chicago Millennium P...     NEUTRAL   
115  Siempre que decidí quedarme en el Fairmont Chi...    POSITIVO   
117  Recientemente me alojé en el hotel Fairmont Ch...    POSITIVO   
163  Me alojé en el Homewood Suites en el centro de...     NEUTRAL   
166  Fui al Homewood Suites en Chicago, que forma p...    POSITIVO   
168  Mi estancia en el Hilton fue como vivir en el ...     NEUTRAL   
169  Al entrar en Homewood Suites, fui recibido por...     NEUTRAL   
171  El Homewood Suites by Hilton, en el centro de ...     NEUTRAL   
172  Recientemente me alojé en el Homewood Suites b...    POSITIVO   
174  Dios mío, este tiene que ser uno de los peores...    POSITIVO   
176  ¡El Homewood Suites es uno de mis hoteles favo...     NEUTRAL   
178  Después de considerar varios hoteles en el áre...    POSITIVO   
180  El hotel Hyatt Regency de Chicago fue el peor ...     NEUTRAL   
181  Durante mi reciente estadía en el Hyatt Regenc...    POSITIVO   
185  Cuando revisé por primera vez el sitio web y l...     NEUTRAL   
187  El Hyatt Regency Chicago parecía un buen lugar...     NEUTRAL   
190  Recientemente me alojé en el hotel Hyatt Regen...     NEUTRAL   
191  Mi esposa y yo fuimos a Chicago para disfrutar...     NEUTRAL   
192  Hotel Hyatt Regency: buen centro de Chicago. L...     NEUTRAL   
194  Mi esposa y yo nos hospedamos en el Hyatt Rege...    NEGATIVO   
196  Decidimos gastar un poco más para alojarnos en...    POSITIVO   
198  Hace poco me alojé en el Hyatt Regency en Chic...     NEUTRAL   
501  Los hoteles Fairmont son generalmente mi opció...    POSITIVO   
504  Tenía muchas ganas de quedarme aquí para mi bo...     NEUTRAL   
505  Hermoso hotel, pero no tiene mucho que ofrecer...     NEUTRAL   
507  El ho

In [45]:
import csv

with open('df_comentarios_negativos.csv', 'w', newline='', encoding='utf-8') as csvfile:
    writer = csv.writer(csvfile)
    writer.writerow(df_comentarios_negativos.columns)  # Escribir encabezados
    writer.writerows(df_comentarios_negativos.values.tolist())  # Escribir filas


In [83]:
import plotly.express as px
import pandas as pd

# Supongamos que ya tienes df_comentarios_negativos con la columna Categoria_Problema

# Contar la cantidad de ocurrencias de cada categoría
categoria_counts = df_comentarios_negativos['Categoria_Problema'].value_counts().reset_index()
categoria_counts.columns = ['Categoria_Problema', 'Count']

# Ordenar por cuenta descendente (opcional)
categoria_counts = categoria_counts.sort_values(by='Count', ascending=False)

# Crear el gráfico de pastel interactivo con plotly
fig = px.pie(categoria_counts, values='Count', names='Categoria_Problema',
             title='Distribución de Problemas en Comentarios Negativos en nuestro hotel',
             color_discrete_sequence=px.colors.qualitative.Set3)

# Mostrar el gráfico
fig.show()


In [97]:
categoria_counts = df_sentimientos_competencia_negativostop['Categoria_Problema'].value_counts().reset_index()
categoria_counts.columns = ['Categoria_Problema', 'Count']

# Ordenar por cuenta descendente (opcional)
categoria_counts = categoria_counts.sort_values(by='Count', ascending=False)

# Crear el gráfico de pastel interactivo con plotly
fig = px.pie(categoria_counts, values='Count', names='Categoria_Problema',
             title='Distribución de Problemas en Comentarios Negativos en los competidores',
             color_discrete_sequence=px.colors.qualitative.Set3)

# Mostrar el gráfico
fig.show()


In [67]:
#Evaluación de la satisfacción del cliente
df_comentarios_positivos=df_sentimientos_hotel[df_sentimientos_hotel['SENTIMIENTO_CALCULADO']=='POSITIVO']
df_comentarios_positivos

Unnamed: 0.1  Unnamed: 0   
0                 0           0  \
3                 3           3   
4                 4           4   
5                 5           5   
6                 6           6   
...             ...         ...   
13125         13125       13125   
13126         13126       13126   
13127         13127       13127   
13128         13128       13128   
13131         13131       13131   

                                       Translated_Review  Rating Sentimiento   
0      bonito hotel, aparcamiento caro, consiguió una...       4    POSITIVO  \
3      Único, estancia estupenda, tiempo maravilloso ...       5     NEUTRAL   
4      estancia estupenda estancia estupenda, fue inc...       5     NEUTRAL   
5      amor personal de mónaco esposo se quedó en el ...       5    POSITIVO   
6      Estancia acogedora en una ciudad lluviosa. Mi ...       5     NEUTRAL   
...                                                  ...     ...         ...   
13125  Novia impresionada, buen ambiente, lugar a un ...       5     NEUTRAL   
13126  buena relación calidad-precio recién devuelta ...       5    NEGATIVO   
13127  Los viajeros caribeños más agradables y experi...       5    POSITIVO   
13128  maravilloso club de vacaciones diamante bahia ...       5    POSITIVO   
13131  excelente piscina, clima increíble, estadía fa...       3    POSITIVO   

      Sentimiento_VADER SENTIMIENTO_CALCULADO  
0              NEGATIVO              POSITIVO  
3              NEGATIVO              POSITIVO  
4              NEGATIVO              POSITIVO  
5              POSITIVO              POSITIVO  
6              NEGATIVO              POSITIVO  
...                 ...                   ...  
13125          NEGATIVO              POSITIVO  
13126          NEGATIVO              POSITIVO  
13127          NEGATIVO              POSITIVO  
13128          NEGATIVO              POSITIVO  
13131          POSITIVO              POSITIVO  

[7417 rows x 7 columns]

In [70]:
import nltk
from nltk.corpus import stopwords

# Descargar los recursos necesarios de NLTK (esto solo se necesita una vez)
nltk.download('stopwords')


categorias_satisfaccion = {
    'Calidad del servicio': [
        'servicio', 'atención', 'amabilidad', 'respuesta', 'trato', 'recepción', 'ayuda', 'cliente',
        'información', 'satisfacción', 'cordialidad', 'profesionalismo', 'hospitalidad', 'cortesía',
        'asesoría', 'asistencia', 'calidad', 'recomendación', 'integridad', 'paciencia', 'dedicación',
        'actitud', 'prestigio', 'garantía', 'compañía', 'solución', 'experiencia', 'estancia', 'momento',
        'viaje', 'opinión', 'satisfacción', 'comentario', 'experiencia'
    ],
    'Calidad de las instalaciones': [
        'instalación', 'limpieza', 'mantenimiento', 'condiciones', 'estado', 'aseo', 'higiene', 'confort',
        'infraestructura', 'servicios', 'ambientación', 'edificio', 'hospitalidad', 'mobiliario', 'comodidad',
        'apariencia', 'materiales', 'estructura', 'diseño', 'construcción', 'decoración', 'comodidades', 'residencia',
        'acabados', 'vivienda', 'casa', 'domicilio', 'detalle', 'calidad', 'planta', 'espacio', 'seguridad', 'oferta',
        'higiene','iluminacion','ruido'
    ],
    'Experiencia general': [
        'experiencia', 'estancia', 'visita', 'viaje', 'impresiones', 'opinión', 'satisfacción', 'comentario',
        'momento', 'estudio', 'reacción', 'recuerdo', 'instante', 'contemplación', 'comunicación', 'tarea',
        'hábito', 'desarrollo', 'condición', 'impacto', 'presente', 'concepto', 'institución', 'ver',
        'formación', 'compromiso', 'pasión', 'carácter', 'comentario', 'instante'
    ],
    'Precios y valor': [
        'precio', 'tarifa', 'costo', 'elevado', 'presupuesto', 'económico', 'costoso', 'gasto',
        'inversión', 'incremento', 'subida', 'alza', 'aumento', 'desembolso', 'desembolso', 'cantidad',
        'nivel', 'monto', 'característica', 'propiedad', 'importe', 'grado', 'escalón', 'tasa', 'pago',
        'especie', 'coste', 'concepto', 'plus', 'cupo', 'clase', 'derecho', 'cuantía', 'cuota', 'chance',
        'impuesto', 'precio', 'tiquete', 'grava', 'desglose', 'valla', 'impresión', 'adición', 'deuda',
        'coste', 'oferta', 'estimación', 'costo', 'valor', 'coste', 'impuesto', 'tasa', 'inflación', 'caída',
        'preocupación', 'billete', 'cambio', 'interés', 'duda', 'descuento', 'gobierno', 'moneda', 'resto',
        'distribución', 'compra', 'futuro', 'asociación', 'parte', 'regulación', 'decreto', 'financiación',
        'subvención', 'planificación', 'periodo', 'economía', 'sistema', 'actuación', 'costo', 'barra', 'tienda',
        'inmueble', 'financiamiento', 'inversión', 'dinero', 'valoración', 'tarifa', 'distribución', 'prueba',
        'aplicación', 'importación', 'precio', 'retiro', 'curso', 'alma', 'grano', 'cultura', 'razón', 'carro',
        'metodología', 'precio','ticket','elevado','carísimo'
    ]
}







[nltk_data] Downloading package stopwords to
[nltk_data]     e:\Users\1077730\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [68]:
import nltk
from collections import defaultdict

stop_words = set(nltk.corpus.stopwords.words('spanish'))

def etiquetar_satisfaccion_nltk(comentario):
    tokens = nltk.word_tokenize(comentario.lower())
    categoria_count = defaultdict(int)

    for categoria, palabras_clave in categorias_satisfaccion.items():
        for palabra_clave in palabras_clave:
            if palabra_clave in tokens and palabra_clave not in stop_words:
                categoria_count[categoria] += 1

    if categoria_count:
        categoria_max = max(categoria_count, key=categoria_count.get)
        return categoria_max
    else:
        return 'No clasificado'


In [71]:
df_comentarios_positivos['Categoria_Satisfaccion'] = df_comentarios_positivos['Translated_Review'].apply(etiquetar_satisfaccion_nltk)
df_comentarios_positivos

e:\Users\1077730\AppData\Local\Temp\ipykernel_24372\4234158082.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Unnamed: 0.1  Unnamed: 0   
0                 0           0  \
3                 3           3   
4                 4           4   
5                 5           5   
6                 6           6   
...             ...         ...   
13125         13125       13125   
13126         13126       13126   
13127         13127       13127   
13128         13128       13128   
13131         13131       13131   

                                       Translated_Review  Rating Sentimiento   
0      bonito hotel, aparcamiento caro, consiguió una...       4    POSITIVO  \
3      Único, estancia estupenda, tiempo maravilloso ...       5     NEUTRAL   
4      estancia estupenda estancia estupenda, fue inc...       5     NEUTRAL   
5      amor personal de mónaco esposo se quedó en el ...       5    POSITIVO   
6      Estancia acogedora en una ciudad lluviosa. Mi ...       5     NEUTRAL   
...                                                  ...     ...         ...   
13125  Novia impresionada, buen ambiente, lugar a un ...       5     NEUTRAL   
13126  buena relación calidad-precio recién devuelta ...       5    NEGATIVO   
13127  Los viajeros caribeños más agradables y experi...       5    POSITIVO   
13128  maravilloso club de vacaciones diamante bahia ...       5    POSITIVO   
13131  excelente piscina, clima increíble, estadía fa...       3    POSITIVO   

      Sentimiento_VADER SENTIMIENTO_CALCULADO Categoria_Satisfaccion  
0              NEGATIVO              POSITIVO   Calidad del servicio  
3              NEGATIVO              POSITIVO   Calidad del servicio  
4              NEGATIVO              POSITIVO        Precios y valor  
5              POSITIVO              POSITIVO   Calidad del servicio  
6              NEGATIVO              POSITIVO   Calidad del servicio  
...                 ...                   ...                    ...  
13125          NEGATIVO              POSITIVO        Precios y valor  
13126          NEGATIVO              POSITIVO        Precios y valor  
13127          NEGATIVO              POSITIVO   Calidad del servicio  
13128          NEGATIVO              POSITIVO   Calidad del servicio  
13131          POSITIVO              POSITIVO        Precios y valor  

[7417 rows x 8 columns]

In [98]:


df_sentimientos_competencia_positivostop['Categoria_Satisfaccion'] = df_sentimientos_competencia_positivostop['Comentario'].apply(etiquetar_satisfaccion_nltk)
df_sentimientos_competencia_positivostop

e:\Users\1077730\AppData\Local\Temp\ipykernel_24372\1824423407.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Unnamed: 0 Nombre Hotel   
340         340      sofitel  \
342         342      sofitel   
343         343      sofitel   
344         344      sofitel   
345         345      sofitel   
347         347      sofitel   
350         350      sofitel   
355         355      sofitel   
359         359      sofitel   
380         380      talbott   
381         381      talbott   
382         382      talbott   
383         383      talbott   
384         384      talbott   
385         385      talbott   
386         386      talbott   
387         387      talbott   
388         388      talbott   
389         389      talbott   
390         390      talbott   
391         391      talbott   
392         392      talbott   
393         393      talbott   
394         394      talbott   
395         395      talbott   
397         397      talbott   
399         399      talbott   
741         741      sofitel   
742         742      sofitel   
743         743      sofitel   
745         745      sofitel   
747         747      sofitel   
748         748      sofitel   
749         749      sofitel   
750         750      sofitel   
751         751      sofitel   
753         753      sofitel   
754         754      sofitel   
755         755      sofitel   
756         756      sofitel   
757         757      sofitel   
758         758      sofitel   
759         759      sofitel   
780         780      talbott   
783         783      talbott   
785         785      talbott   
786         786      talbott   
788         788      talbott   
790         790      talbott   
791         791      talbott   
792         792      talbott   
795         795      talbott   
796         796      talbott   
797         797      talbott   
798         798      talbott   

                                            Comentario Sentimiento   
340  Necesitaba un lugar donde quedarme para una co...    NEGATIVO  \
342  Me alojé en Sofitel con mi marido durante un f...     NEUTRAL   
343  Mi marido y yo nos alojamos en el Sofitel Chic...     NEUTRAL   
344  Alojarme en el Sofitel fue una de las experien...     NEUTRAL   
345  Después de llegar al hotel Sofitel Chicago Wat...     NEUTRAL   
347  El Sofitel Chicago Water Tower, aunque en teor...    POSITIVO   
350  Me alojé en este hotel por 2 noches. Tenía gra...     NEUTRAL   
355  Hace poco me alojé en el hotel Sofitel (Chicag...    POSITIVO   
359  Estaba aquí por negocios, así que necesitaba d...     NEUTRAL   
380  Recientemente me alojé en el Hotel Talbott en ...     NEUTRAL   
381  "Con clase", "Decente" y "Serena". Estas palab...     NEUTRAL   
382  ¡El Hotel Talbott es fantástico! Me quedé allí...     NEUTRAL   
383  Este hotel era todo lo que estábamos buscando....     NEUTRAL   
384  Mi marido y yo decidimos planear una escapada ...     NEUTRAL   
385  Este es el mejor hotel para alojarse cuando vi...     NEUTRAL   
386  Mi estadía en el Talbott fue una experiencia m...     NEUTRAL   
387  ¡Gran hotel! El personal es amable, cortés y p...     NEUTRAL   
388  Realmente me encantó mi estancia en el Talbott...     NEUTRAL   
389  Este hotel fue un auténtico sueño en el que al...     NEUTRAL   
390  Muy lujoso y vale la pena el dinero. Si busca ...     NEUTRAL   
391  Estaba visitando a unos familiares en Chicago ...    POSITIVO   
392  El hotel Talbot es un lugar elegante para llev...     NEUTRAL   
393  El ambiente en el Talbott Hotel es muy acogedo...    POSITIVO   
394  Al principio, mi esposo y yo estábamos muy emo...    POSITIVO   
395  Las tarifas en The Talbott Hotel eran más bara...    POSITIVO   
397  Personal agradable y servicio de limpieza. Des...     NEUTRAL   
399  Excelente personal y servicio al cliente, muy ...     NEUTRAL   
741  Retiré mi Lexus y me lo devolvieron con el esp...     NEUTRAL   
742  Tuve la ceremonia de mi boda y la recepción en...    POSITIVO   
743  Leí las críticas en Trip Advisor sobre esta pr...     NEUTRAL   
745  Acabo de regresar de una visita de 10

In [73]:
import plotly.express as px
import pandas as pd

# Supongamos que ya tienes df_comentarios_negativos con la columna Categoria_Problema

# Contar la cantidad de ocurrencias de cada categoría
categoria_counts = df_comentarios_positivos['Categoria_Satisfaccion'].value_counts().reset_index()
categoria_counts.columns = ['Categoria_Satisfaccion', 'Count']

# Ordenar por cuenta descendente (opcional)
categoria_counts = categoria_counts.sort_values(by='Count', ascending=False)

# Crear el gráfico de pastel interactivo con plotly
fig = px.pie(categoria_counts, values='Count', names='Categoria_Satisfaccion',
             title='Distribución de categorías de satisfacción del cliente',
             color_discrete_sequence=px.colors.qualitative.Set3)

# Mostrar el gráfico
fig.show()


In [99]:
import plotly.express as px
import pandas as pd

categoria_counts = df_sentimientos_competencia_positivostop['Categoria_Satisfaccion'].value_counts().reset_index()
categoria_counts.columns = ['Categoria_Satisfaccion', 'Count']

# Ordenar por cuenta descendente (opcional)
categoria_counts = categoria_counts.sort_values(by='Count', ascending=False)

# Crear el gráfico de pastel interactivo con plotly
fig = px.pie(categoria_counts, values='Count', names='Categoria_Satisfaccion',
             title='Distribución de categorías de satisfacción del cliente',
             color_discrete_sequence=px.colors.qualitative.Set3)

# Mostrar el gráfico
fig.show()



In [106]:
df_sentimientos_hotel

Unnamed: 0.1  Unnamed: 0   
0                 0           0  \
1                 1           1   
2                 2           2   
3                 3           3   
4                 4           4   
...             ...         ...   
13127         13127       13127   
13128         13128       13128   
13129         13129       13129   
13130         13130       13130   
13131         13131       13131   

                                       Translated_Review  Rating Sentimiento   
0      bonito hotel, aparcamiento caro, consiguió una...       4    POSITIVO  \
1      ok, nada de cargo especial, miembro diamante h...       2    POSITIVO   
2      Bonitas habitaciones, no un hotel de 4* con ex...       3    NEGATIVO   
3      Único, estancia estupenda, tiempo maravilloso ...       5     NEUTRAL   
4      estancia estupenda estancia estupenda, fue inc...       5     NEUTRAL   
...                                                  ...     ...         ...   
13127  Los viajeros caribeños más agradables y experi...       5    POSITIVO   
13128  maravilloso club de vacaciones diamante bahia ...       5    POSITIVO   
13129  No me molesté en quedarme en la puerta, me que...       1    NEGATIVO   
13130  escribió el encabezado, breve. comida repugnan...       1     NEUTRAL   
13131  excelente piscina, clima increíble, estadía fa...       3    POSITIVO   

      Sentimiento_VADER SENTIMIENTO_CALCULADO  
0              NEGATIVO              POSITIVO  
1              NEGATIVO               NEUTRAL  
2              NEGATIVO              NEGATIVO  
3              NEGATIVO              POSITIVO  
4              NEGATIVO              POSITIVO  
...                 ...                   ...  
13127          NEGATIVO              POSITIVO  
13128          NEGATIVO              POSITIVO  
13129          NEGATIVO              NEGATIVO  
13130           NEUTRAL              NEGATIVO  
13131          POSITIVO              POSITIVO  

[13132 rows x 7 columns]

In [119]:
df_sentimientos_hotel

Unnamed: 0.1  Unnamed: 0   
0                 0           0  \
1                 1           1   
2                 2           2   
3                 3           3   
4                 4           4   
...             ...         ...   
13127         13127       13127   
13128         13128       13128   
13129         13129       13129   
13130         13130       13130   
13131         13131       13131   

                                       Translated_Review  Rating Sentimiento   
0      bonito hotel, aparcamiento caro, consiguió una...       4    POSITIVO  \
1      ok, nada de cargo especial, miembro diamante h...       2    POSITIVO   
2      Bonitas habitaciones, no un hotel de 4* con ex...       3    NEGATIVO   
3      Único, estancia estupenda, tiempo maravilloso ...       5     NEUTRAL   
4      estancia estupenda estancia estupenda, fue inc...       5     NEUTRAL   
...                                                  ...     ...         ...   
13127  Los viajeros caribeños más agradables y experi...       5    POSITIVO   
13128  maravilloso club de vacaciones diamante bahia ...       5    POSITIVO   
13129  No me molesté en quedarme en la puerta, me que...       1    NEGATIVO   
13130  escribió el encabezado, breve. comida repugnan...       1     NEUTRAL   
13131  excelente piscina, clima increíble, estadía fa...       3    POSITIVO   

      Sentimiento_VADER SENTIMIENTO_CALCULADO            Categoria_Problema  
0              NEGATIVO              POSITIVO  Calidad de las instalaciones  
1              NEGATIVO               NEUTRAL  Calidad de las instalaciones  
2              NEGATIVO              NEGATIVO   Precio y costos adicionales  
3              NEGATIVO              POSITIVO           Servicio al cliente  
4              NEGATIVO              POSITIVO  Calidad de las instalaciones  
...                 ...                   ...                           ...  
13127          NEGATIVO              POSITIVO           Servicio al cliente  
13128          NEGATIVO              POSITIVO  Calidad de las instalaciones  
13129          NEGATIVO              NEGATIVO           Servicio al cliente  
13130           NEUTRAL              NEGATIVO  Calidad de las instalaciones  
13131          POSITIVO              POSITIVO   Precio y costos adicionales  

[13132 rows x 8 columns]

In [120]:
#MODELO
df_sentimientos_hotel_filt=df_sentimientos_hotel[['Translated_Review','SENTIMIENTO_CALCULADO','Categoria_Problema']]
df_sentimientos_hotel_filt['Hotel'] = 'Hotel cliente'
df_sentimientos_hotel_filt

e:\Users\1077730\AppData\Local\Temp\ipykernel_24372\353272050.py:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Translated_Review   
0      bonito hotel, aparcamiento caro, consiguió una...  \
1      ok, nada de cargo especial, miembro diamante h...   
2      Bonitas habitaciones, no un hotel de 4* con ex...   
3      Único, estancia estupenda, tiempo maravilloso ...   
4      estancia estupenda estancia estupenda, fue inc...   
...                                                  ...   
13127  Los viajeros caribeños más agradables y experi...   
13128  maravilloso club de vacaciones diamante bahia ...   
13129  No me molesté en quedarme en la puerta, me que...   
13130  escribió el encabezado, breve. comida repugnan...   
13131  excelente piscina, clima increíble, estadía fa...   

      SENTIMIENTO_CALCULADO            Categoria_Problema          Hotel  
0                  POSITIVO  Calidad de las instalaciones  Hotel cliente  
1                   NEUTRAL  Calidad de las instalaciones  Hotel cliente  
2                  NEGATIVO   Precio y costos adicionales  Hotel cliente  
3                  POSITIVO           Servicio al cliente  Hotel cliente  
4                  POSITIVO  Calidad de las instalaciones  Hotel cliente  
...                     ...                           ...            ...  
13127              POSITIVO           Servicio al cliente  Hotel cliente  
13128              POSITIVO  Calidad de las instalaciones  Hotel cliente  
13129              NEGATIVO           Servicio al cliente  Hotel cliente  
13130              NEGATIVO  Calidad de las instalaciones  Hotel cliente  
13131              POSITIVO   Precio y costos adicionales  Hotel cliente  

[13132 rows x 4 columns]

In [121]:
df_sentimientos_competencia_filt=df_sentimientos_competencia[['Comentario','Sentimiento_VADER','Categoria_Problema','Nombre Hotel']]
df_sentimientos_competencia_filt

Comentario Sentimiento_VADER   
0    Después de una estancia de una semana reciente...          POSITIVO  \
1    Aunque demasiado caro en mi opinión, el hotel ...          POSITIVO   
2    Me alojé en el Affina Chicago para celebrar mi...          POSITIVO   
3    ESTE HOTEL ES FANTÁSTICO. Me alojé allí en mi ...          POSITIVO   
4    El Affinia Chicago es un lugar maravilloso don...          POSITIVO   
..                                                 ...               ...   
795  Me alojé en The Talbott por 3 noches por negoc...          POSITIVO   
796  Nos encantó el Talbott. La ubicación era fabul...          POSITIVO   
797  Puedo decir fácilmente que este es uno de los ...          POSITIVO   
798  Este es un gran hallazgo para el centro de Chi...          POSITIVO   
799  Chicago es una de nuestras ciudades favoritas ...           NEUTRAL   

               Categoria_Problema Nombre Hotel  
0    Calidad de las instalaciones      affinia  
1             Servicio al cliente      affinia  
2             Servicio al cliente      affinia  
3             Servicio al cliente      affinia  
4    Calidad de las instalaciones      affinia  
..                            ...          ...  
795  Calidad de las instalaciones      talbott  
796  Calidad de las instalaciones      talbott  
797  Calidad de las instalaciones      talbott  
798   Precio y costos adicionales      talbott  
799           Servicio al cliente      talbott  

[800 rows x 4 columns]

In [122]:
# Renombrar columnas en df_sentimientos_competencia_filt para que coincidan con df_sentimientos_hotel_filt
df_sentimientos_competencia_filt.rename(columns={'Comentario': 'Translated_Review',
                                                'Sentimiento_VADER': 'SENTIMIENTO_CALCULADO',
                                                'Nombre Hotel': 'Hotel'}, inplace=True)


e:\Users\1077730\AppData\Local\Temp\ipykernel_24372\607195323.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [123]:
import pandas as pd

# Concatenar los dos DataFrames
df_concatenado = pd.concat([df_sentimientos_hotel_filt, df_sentimientos_competencia_filt], ignore_index=True)

df_concatenado

Translated_Review   
0      bonito hotel, aparcamiento caro, consiguió una...  \
1      ok, nada de cargo especial, miembro diamante h...   
2      Bonitas habitaciones, no un hotel de 4* con ex...   
3      Único, estancia estupenda, tiempo maravilloso ...   
4      estancia estupenda estancia estupenda, fue inc...   
...                                                  ...   
13927  Me alojé en The Talbott por 3 noches por negoc...   
13928  Nos encantó el Talbott. La ubicación era fabul...   
13929  Puedo decir fácilmente que este es uno de los ...   
13930  Este es un gran hallazgo para el centro de Chi...   
13931  Chicago es una de nuestras ciudades favoritas ...   

      SENTIMIENTO_CALCULADO            Categoria_Problema          Hotel  
0                  POSITIVO  Calidad de las instalaciones  Hotel cliente  
1                   NEUTRAL  Calidad de las instalaciones  Hotel cliente  
2                  NEGATIVO   Precio y costos adicionales  Hotel cliente  
3                  POSITIVO           Servicio al cliente  Hotel cliente  
4                  POSITIVO  Calidad de las instalaciones  Hotel cliente  
...                     ...                           ...            ...  
13927              POSITIVO  Calidad de las instalaciones        talbott  
13928              POSITIVO  Calidad de las instalaciones        talbott  
13929              POSITIVO  Calidad de las instalaciones        talbott  
13930              POSITIVO   Precio y costos adicionales        talbott  
13931               NEUTRAL           Servicio al cliente        talbott  

[13932 rows x 4 columns]

In [128]:
# Codificar la columna Sentimiento como variables numéricas (ejemplo)
df_concatenado['Sentimiento_numerico'] = df_concatenado['SENTIMIENTO_CALCULADO'].map({'POSITIVO': 1, 'NEGATIVO': -1, 'NEUTRAL': 0})

# Codificar la columna Categoria_Satisfaccion como variables numéricas (ejemplo)
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
df_concatenado['Categoria_Problema_numerica'] = encoder.fit_transform(df_concatenado['Categoria_Problema'])


In [129]:
df_concatenado

Translated_Review   
0      bonito hotel, aparcamiento caro, consiguió una...  \
1      ok, nada de cargo especial, miembro diamante h...   
2      Bonitas habitaciones, no un hotel de 4* con ex...   
3      Único, estancia estupenda, tiempo maravilloso ...   
4      estancia estupenda estancia estupenda, fue inc...   
...                                                  ...   
13927  Me alojé en The Talbott por 3 noches por negoc...   
13928  Nos encantó el Talbott. La ubicación era fabul...   
13929  Puedo decir fácilmente que este es uno de los ...   
13930  Este es un gran hallazgo para el centro de Chi...   
13931  Chicago es una de nuestras ciudades favoritas ...   

      SENTIMIENTO_CALCULADO            Categoria_Problema          Hotel   
0                  POSITIVO  Calidad de las instalaciones  Hotel cliente  \
1                   NEUTRAL  Calidad de las instalaciones  Hotel cliente   
2                  NEGATIVO   Precio y costos adicionales  Hotel cliente   
3                  POSITIVO           Servicio al cliente  Hotel cliente   
4                  POSITIVO  Calidad de las instalaciones  Hotel cliente   
...                     ...                           ...            ...   
13927              POSITIVO  Calidad de las instalaciones        talbott   
13928              POSITIVO  Calidad de las instalaciones        talbott   
13929              POSITIVO  Calidad de las instalaciones        talbott   
13930              POSITIVO   Precio y costos adicionales        talbott   
13931               NEUTRAL           Servicio al cliente        talbott   

       Sentimiento_numerico  Categoria_Problema_numerica  
0                         1                            0  
1                         0                            0  
2                        -1                            2  
3                         1                            3  
4                         1                            0  
...                     ...                          ...  
13927                     1                            0  
13928                     1                            0  
13929                     1                            0  
13930                     1                            2  
13931                     0                            3  

[13932 rows x 6 columns]

In [142]:
# Seleccionar características
X = df_concatenado[['Translated_Review', 'SENTIMIENTO_CALCULADO', 'Sentimiento_numerico', 'Categoria_Problema_numerica']]

# Definir la variable objetivo
y = df_concatenado['Categoria_Problema_numerica']  


In [147]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline

# Ejemplo de datos de entrada (asegúrate de usar tus propios datos aquí)
X_train = df_concatenado['Translated_Review']  # Aquí debes usar la columna que contiene los comentarios
y_train = df_concatenado['Categoria_Problema']

# Dividir datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X_train, y_train, test_size=0.2, random_state=42)

# Definir un pipeline para procesamiento de texto y clasificación
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer()),  # Vectorización de texto
    ('clf', LogisticRegression()),  # Clasificador (puedes usar otro modelo aquí)
])

# Ajustar el pipeline al conjunto de entrenamiento
pipeline.fit(X_train, y_train)

# Evaluar el pipeline en el conjunto de prueba
accuracy = pipeline.score(X_test, y_test)
print(f'Exactitud del modelo: {accuracy}')


e:\Desarrollos\lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning:

lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression



Exactitud del modelo: 0.7527807678507356


In [150]:
# Predicción
y_pred = pipeline.predict(X_test)

# Evaluación del modelo
from sklearn.metrics import classification_report, accuracy_score

print('Exactitud:', accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))


Exactitud: 0.7527807678507356
                              precision    recall  f1-score   support

Calidad de las instalaciones       0.73      0.87      0.80      1237
         Experiencia general       0.00      0.00      0.00        77
 Precio y costos adicionales       0.87      0.50      0.63       504
         Servicio al cliente       0.75      0.80      0.77       969

                    accuracy                           0.75      2787
                   macro avg       0.59      0.54      0.55      2787
                weighted avg       0.74      0.75      0.74      2787



e:\Desarrollos\lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning:

Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.

e:\Desarrollos\lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning:

Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.

e:\Desarrollos\lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning:

Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.



Exactitud

La exactitud del modelo es del 75.28%, lo que indica que en general, el 75.28% de las predicciones hechas por el modelo son correctas en el conjunto de datos de prueba.
Informe de clasificación

El informe de clasificación proporciona precision, recall (también conocido como sensibilidad) y f1-score para cada clase del problema:

    Calidad de las instalaciones:
        Precision: 0.73
        Recall: 0.87
        F1-score: 0.80
        Support: 1237

    Esto indica que el modelo predijo correctamente el 73% de las instancias etiquetadas como "Calidad de las instalaciones". Además, el recall es alto (87%), lo que sugiere que el modelo identifica la mayoría de las instancias relevantes para esta clase.

    Experiencia general:
        Precision: 0.00
        Recall: 0.00
        F1-score: 0.00
        Support: 77

    Para la clase "Experiencia general", el modelo no predijo correctamente ninguna instancia. Esto podría indicar un desequilibrio de clases o dificultades específicas con esta categoría.

    Precio y costos adicionales:
        Precision: 0.87
        Recall: 0.50
        F1-score: 0.63
        Support: 504

    El modelo tiene una alta precision (87%) para predecir la categoría "Precio y costos adicionales", pero el recall es menor (50%), lo que sugiere que no identifica todas las instancias relevantes para esta clase.

    Servicio al cliente:
        Precision: 0.75
        Recall: 0.80
        F1-score: 0.77
        Support: 969

    Para la clase "Servicio al cliente", el modelo tiene un buen balance entre precision (75%) y recall (80%), lo que indica que predice correctamente la mayoría de las instancias positivas y no predice demasiados falsos positivos.

Análisis

    General: La precisión y el recall varían significativamente entre las diferentes categorías, lo que sugiere que el modelo puede tener un desempeño desigual dependiendo de la clase.
    Experiencia general: La falta de predicciones correctas para esta clase específica indica que podría ser beneficioso explorar por qué el modelo tiene dificultades con esta categoría en particular.
    Precio y costos adicionales: Aunque la precisión es alta, el recall es bajo, lo que sugiere que el modelo podría beneficiarse de ajustes para capturar más instancias de esta categoría.

Aumento de datos: Si hay clases con pocos datos (como "Experiencia general"), recopilar más datos o utilizar técnicas de aumento de datos podría ayudar.
Ajuste de hiperparámetros: Experimentar con diferentes configuraciones de hiperparámetros para el vectorizador TF-IDF y el clasificador (por ejemplo, regularización en la regresión logística) podría mejorar el rendimiento.
Exploración de características: Revisar las características más importantes o la importancia de las palabras/claves podría proporcionar información sobre cómo el modelo toma decisiones.

In [ ]:
# Supongamos que tienes un nuevo conjunto de datos de comentarios a predecir en df_nuevos_comentarios
X_nuevos = df_nuevos_comentarios['Translated_Review']

# Aplica el mismo preprocesamiento que hiciste en X_train
X_nuevos_procesados = pipeline.named_steps['tfidf'].transform(X_nuevos)


In [ ]:
# Predicciones
y_pred_nuevos = pipeline.predict(X_nuevos_procesados)

# Añadir las predicciones al DataFrame original
df_nuevos_comentarios['Categoria_Predicha'] = y_pred_nuevos


Exactitud general del modelo: El modelo logra una exactitud global del 75.28%, lo cual indica que es capaz de predecir correctamente las categorías de satisfacción o problemas en aproximadamente tres cuartas partes de los casos.

Precisión por categoría:

    La categoría Calidad de las instalaciones tiene una precisión del 73%, lo cual sugiere que el modelo es relativamente bueno identificando comentarios relacionados con la calidad de las instalaciones.
    La categoría Servicio al cliente presenta una precisión del 75%, indicando que también es acertado en clasificar comentarios que mencionan aspectos del servicio al cliente.
    La categoría Precio y costos adicionales tiene una precisión del 87%, lo que sugiere una buena capacidad del modelo para identificar comentarios sobre aspectos económicos y de costos adicionales.
    Sin embargo, la categoría Experiencia general muestra un rendimiento pobre con un f1-score de 0.00, lo cual indica que el modelo no está prediciendo bien esta categoría específica. Esto puede deberse a un desbalance en los datos o a una complejidad en la variabilidad de los comentarios que no está siendo captada adecuadamente por el modelo.

Macro avg y weighted avg: El macro avg y weighted avg para precision, recall y f1-score muestran que el modelo tiene un rendimiento moderado en promedio, con ciertas categorías destacando más que otras.

Conclusión general: El modelo demuestra ser útil para predecir categorías específicas de satisfacción o problemas en comentarios de clientes de hoteles, especialmente en lo que respecta a la calidad de las instalaciones, servicio al cliente y precios/costos adicionales. Sin embargo, existe margen para mejorar la capacidad del modelo para clasificar correctamente la experiencia general, lo cual es crucial para obtener una visión completa de la satisfacción del cliente.

Recomendaciones: Es recomendable realizar análisis adicionales para entender mejor las limitaciones del modelo en la categoría de experiencia general y considerar ajustes en el preprocesamiento de texto, la selección del modelo o la inclusión de más datos para mejorar la precisión general del modelo.